### PR TITLE
os/FileStore: fix fiemap issue in xfs when #extents > 1364 

### PR DESCRIPTION
**Pull request by mslovy**
fiemap extents cannot be totally retrieved when the number of extents > 1364

### CODE REVIEW

**Comment by tchaikov** [comment]
tested at http://pulpito.ceph.com/kchai-2016-10-24_03:13:40-rados-wip-kefu-testing---basic-mira/

**Comment by mslovy** [comment]
hi, @tchaikov ,  24 failed and 1 dead in pulpito qa test.  It seems the log is quite large,  is any Keywords I can search to quickly find where  the error occur in the log file?

**Comment by tchaikov** [comment]
@mslovy the test passed. the failed ones are irrelevant.

File: src/os/filestore/FileStore.cc [hunk 1]
```diff
@@ -3199,9 +3201,14 @@ int FileStore::_do_fiemap(int fd, uint64_t offset, size_t len,
       extent->fe_length = offset + len - extent->fe_logical;
     (*m)[extent->fe_logical] = extent->fe_length;
     i++;
-    extent++;
+    last = extent++;
   }
   free(fiemap);
+  if (!(last->fe_flags & FIEMAP_EXTENT_LAST)) {
+    offset = last->fe_logical + last->fe_length;
+    len -= offset;
```
Comment (liewegas) [inline src/os/filestore/FileStore.cc]: This len -= line looks wrong... it assumes the original offset was 0, which is isn't (esp if you end up doing 2 batches of fiemaps).

**Comment by mslovy** [comment]
thx, fix it.

File: src/os/filestore/FileStore.cc [hunk 2]
```diff
@@ -3199,9 +3201,15 @@ int FileStore::_do_fiemap(int fd, uint64_t offset, size_t len,
       extent->fe_length = offset + len - extent->fe_logical;
     (*m)[extent->fe_logical] = extent->fe_length;
     i++;
-    extent++;
+    last = extent++;
   }
   free(fiemap);
+  if (!(last->fe_flags & FIEMAP_EXTENT_LAST)) {
+    uint64_t xoffset = last->fe_logical + last->fe_length - offset;
+    offset = last->fe_logical + last->fe_length;
```
Comment (liewegas) [inline src/os/filestore/FileStore.cc]: offset += xoffset;
too, right?

**Comment by smithfarm** [comment]
The hammer backport is non-trivial. Anyone here care to take it on? It's http://tracker.ceph.com/issues/17878

**Comment by smithfarm** [comment]
Ah, nevermind, @mslovy opened #11615

**Comment by smithfarm** [comment]
@mslovy While you're rebasing, please re-do the cherry-pick of 1a1c126d80b427d5230347fbc71a4edea5c0b6c8 using `git cherry-pick -x` -- the "(cherry picked from . . .)" line is needed in the commit message.

**Pull request merged**